File: frontend/src/app/services/conta.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Conta, StatusConta, TipoConta } from '../models/conta.model';
import { environment } from '../../environments/environment';

@Injectable({ providedIn: 'root' })
export class ContaService {

  private readonly apiUrl = environment.apiUrl;

  constructor(private http: HttpClient) {}

  listar(status?: StatusConta, tipo?: TipoConta): Observable<Conta[]> {
    const params: string[] = [];
    if (status) {
      params.push(`status=${status}`);
    }
    if (tipo) {
      params.push(`tipo=${tipo}`);
    }
    const query = params.length ? `?${params.join('&')}` : '';
    return this.http.get<Conta[]>(`${this.apiUrl}${query}`);
  }

  saldoTotal(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/saldo-total`);
  }

  abrir(conta: Conta): Observable<Conta> {
    return this.http.post<Conta>(this.apiUrl, conta);
  }

  atualizar(id: number, conta: Conta): Observable<Conta> {
    return this.http.put<Conta>(`${this.apiUrl}/${id}`, conta);
  }

  encerrar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  depositar(id: number, valor: number): Observable<Conta> {
    return this.http.post<Conta>(`${this.apiUrl}/${id}/depositar`, { valor });
  }

  sacar(id: number, valor: number): Observable<Conta> {
    return this.http.post<Conta>(`${this.apiUrl}/${id}/sacar`, { valor });
  }
}

File: backend/src/main/java/com/entrevista/banco/service/ContaService.java
package com.entrevista.banco.service;

import com.entrevista.banco.domain.StatusConta;
import com.entrevista.banco.domain.TipoConta;
import com.entrevista.banco.dto.ContaRequest;
import com.entrevista.banco.dto.ContaResponse;
import com.entrevista.banco.dto.MovimentoRequest;

import java.util.List;

public interface ContaService {

    ContaResponse abrir(ContaRequest request);

    ContaResponse buscarPorId(Long id);

    List<ContaResponse> listar(StatusConta status, TipoConta tipo);

    ContaResponse atualizar(Long id, ContaRequest request);

    void encerrar(Long id);

    ContaResponse depositar(Long id, MovimentoRequest request, String idempotencyKey);

    ContaResponse sacar(Long id, MovimentoRequest request, String idempotencyKey);
}
